docs/architecture/adr-071-crypto-v2-multi-curve.md

```diff
@@ -0,0 +1,804 @@
+
+This **interface** acts as a centralized controller, encapsulating the APIs for the **signing**, **verifying** and **hashing** functionalities. It acts as the main API with which the apps will interact with
+
+By abstracting the underlying cryptographic functionalities, `CryptoProvider` enables a modular and extensible architecture, aka 'pluggable cryptography'. It allows users to easily switch between different cryptographic implementations without impacting the rest of the system.
+
+The `CryptoProvider` interface includes getters for essential cryptographic functionalities and its metadata:
+
+
+```go
+// CryptoProvider aggregates the functionalities of signing, verifying, and hashing, and provides metadata.
+type CryptoProvider interface {
+    // GetSigner returns an instance of Signer.
+    GetSigner() Signer
+
+    // GetVerifier returns an instance of Verifier.
+    GetVerifier() Verifier
+
+    // GetHasher returns an instance of Hasher.
+    GetHasher() Hasher
+
+    // Metadata returns metadata for the crypto provider.
+    Metadata() ProviderMetadata
+}
+```
+
+##### Components
+
+The components defined here are designed to act as *wrappers* around the underlying proper functions. This architecture ensures that the actual cryptographic operations such as signing, hashing, and verifying are delegated to the specialized functions, that are implementation dependant. These wrapper components facilitate a clean and modular approach by abstracting the complexity of direct cryptographic function calls.
+
+In all of the interface's methods, we add an *options* input parameter of type `map[string]any`, designed to provide a flexible and dynamic way to pass various options and configurations to the `Sign`, `Verify`, and `Hash` functions. This approach allows developers to customize these processes by including any necessary parameters that might be required by specific algorithms or operational contexts. However, this requires that a type assertion for each option be performed inside the function's implementation.
+
+###### Signer
+
+Interface responsible for signing a message and returning the generated signature. 
+The `SignerOptions` map allows for flexible and dynamic configuration of the signing process.
+This can include algorithm-specific parameters, security levels, or other contextual information
+that might be necessary for the signing operation.
+
+```go
+// Signer represents a general interface for signing messages.
+type Signer interface {
+    // Sign takes a signDoc as input and returns the digital signature.
+    Sign(signDoc []byte, options SignerOptions) (Signature, error)
+}
+
+type SignerOpts = map[string]any
+```
+
+###### Signature
+
+```go
+// Signature represents a general interface for a digital signature.
+type Signature interface {
+    // Bytes returns the byte representation of the signature.
+    Bytes() []byte
+
+    // Equals checks if two signatures are identical.
+    Equals(other Signature) bool
```

Comment:
Would we need `Type` here as in `PublicKey` below?

Reply:
yes, that's a good one. Thanks!